### PR TITLE
pkg/mynewt-core: initial commit

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

Currently, two packages that were originally based on mynewt are supported in RIOT `nimble` and `uwb-core`. Both take a similar approach to implementing a porting layer. Some things are properly abstracted but other modules are simply copied over from mynewt. This leads to the issue described in #15528.

This PR attempts to fix that by providing a mynewt-core pkg so that all possible code is shared between both packages and so that it comes from the original pkg. Where possible the api is inlined to wrap around RIOT's api. In other cases, the modules are compiled directly from mynewt.

To avoid useless dependencies and inclusion of unwanted headers I put all the required headers in their own folder in the mynewt-core pkg. This leads to a verbose patch but I couldn't think of a better solution, suggestions are welcomed.

There is one hardware dependency that is already present in nimble and that is kept here and it relates to `os_cputime` that is built on top of `mynewt` `hal_timer` to run on `nrf5x` second rtc. I initially tried to run this on top of ztimer but there is some very special handling going on there that I could not reproduce, maybe @haukepetersen has tried this already and has more insight.

On the nimble side `pkg/nimble/npl/include/nimble/nimble_npl_os.h` and other porting files are now hosted in RIOT. If the pkg maintainer prefers it, this could still be done in nimble, but I don't min going through all the doxygen, and other nitpicks to include in RIOT if that is preferred.

TODO:
- documentation on the package integration
- probably a lot of static-check issues.

@haukepetersen what do you think of this approach? 

### Testing procedure

I verified that:

- `examples/twr_aloha` still work
- `examples/gnrc_networkin` work on `nrf52` boards
- and tested both packages on https://gitlab.inria.fr/pepper/riot-demo-apps/-/tree/app/ble-rng-gatt/03_ble_rng_gatt

@haukepetersen might be able to do some more intese testing?

The `os/%` module abstraction should be checked so that I didn't miss anything that might have been done differently in nimble compared to uwb-core

### Issues/PRs references

Fixes https://github.com/RIOT-OS/RIOT/issues/15528

### CODE REVIEW

**Comment by fjmolinas** [comment]
@rdagher might be interested in this as well.

**Comment by miri64** [comment]
Can we name the package something like `mynewt-adapter` or something like that? When I read the title I was at first confused, why you want to have the mynewt kernel in RIOT :sweat_smile:.

**Comment by fjmolinas** [comment]
> Can we name the package something like `mynewt-adapter` or something like that? When I read the title I was at first confused, why you want to have the mynewt kernel in RIOT sweat_smile

I can yes, although  IMO it's better to use the actual repo name for the pkg name.

**Comment by haukepetersen** [comment]
Now I finally went through this PR in detail, again sorry for the delay.

Of course this is one big-ass change, hard to get the grasp of it :-) The general direction looks sane to me and I ran some of my experiments this morning: so far everything is looking smoothly, that's good!

For the specific changes, I have only looked at nimble so far. I mainly have some general questions/remarks:
1. IMHO it would be great, at least for a transition time, to be able to build NimBLE with both, the NimBLE-included NPL implementation and the new `mynewt-core` package. This would significantly help for bug hunting as we have some means to find out if a faulty behavior is caused by the `mynewt-core` package or if the issue is in the NimBLE code. As discussed offline, I have spend the good part of a year to hunt down bugs in NimBLE and the RIOT NPL implementation, so my upmost concern is to keep the stability we have currently for any means! So this option for regression testing is quite important to me.
2. I am not sure what best to do with the `syscfg.h` files. So far the one for NimBLE is, as it states in the files comments, auto-generated through the NimBLE tool based on NimBLE's `syscfg.yml` files. This ensures, at least for the NimBLE part, that all additions/changes to the default NimBLE configuration are automatically also pulled in RIOT if the NimBLE version is updated. Do you think we could come up with some tooling that for example would take the NimBLE generated file as input and simply remove all the options that are covered by the `mynewt-core` module? 
3. At least for the NimBLE package I strongly want to go without patches, as I personally think they are a maintenance nightmare. Gettiing changes into NimBLE code upstream is no big deal, so adapting whats needed adaption there would be the way to go. But I don't quite have good idea on how to upstream the fix to `nimble_porting.c` in a way, that would allow for both, compiling with the NimBLE NPL or the `mynewt-core`. Probably a RIOT specific configuration option in NimBLE would do the trick?!

File: pkg/mynewt-core/include/os/os.h
Comment (haukepetersen) [inline pkg/mynewt-core/include/os/os.h:16]: not sure, but this file is mostly copied over from "MYNEWT/kernel/os/include/os/os.h", is it not? Could this bring is into trouble regarding licence issues or just marking it as one of our own files?

Comment (haukepetersen) [reply pkg/mynewt-core/include/os/os.h:16]: This might apply for more files in this include folder, right?!

File: pkg/mynewt-core/include/syscfg/syscfg.h
Comment (haukepetersen) [inline pkg/mynewt-core/include/syscfg/syscfg.h:78]: shouldn't we pull these defines to the top of the file to make sure that neither Nimble nor UWB configs override these?

File: pkg/mynewt-core/patches/0001-riot-add-interested-header-symlinks.patch
Comment (haukepetersen) [inline pkg/mynewt-core/patches/0001-riot-add-interested-header-symlinks.patch:7]: do we really need to add these headers via patch? Can't we just put them into their own include dir(s) in `pkg/mynewt-core/...` and use them from there?

**Review by haukepetersen — COMMENTED** [review]
Some more small questions :-)

File: pkg/mynewt-core/patches/0001-riot-add-interested-header-symlinks.patch
Comment (fjmolinas) [reply pkg/mynewt-core/patches/0001-riot-add-interested-header-symlinks.patch:7]: I tried to do a symlink so that the files would stay in sync if ever changed... but it ended up being a huge patch, and I did this to avoid unwanted headers being included. What I wanted to avoid was cluttering the code base with files that are exact copies of files in `mynewt-core`. I'll see if there is another solution I can try.

File: pkg/mynewt-core/include/os/os.h
Comment (fjmolinas) [reply pkg/mynewt-core/include/os/os.h:16]: True... I'll see if I can figure out a less invasive solution...

**Comment by fjmolinas** [comment]
> 1. IMHO it would be great, at least for a transition time, to be able to build NimBLE with both, the NimBLE-included NPL implementation and the new `mynewt-core` package. This would significantly help for bug hunting as we have some means to find out if a faulty behavior is caused by the `mynewt-core` package or if the issue is in the NimBLE code. As discussed offline, I have spend the good part of a year to hunt down bugs in NimBLE and the RIOT NPL implementation, so my upmost concern is to keep the stability we have currently for any means! So this option for regression testing is quite important to me.

That makes sense, I'll take a look at this.

> 2. I am not sure what best to do with the `syscfg.h` files. So far the one for NimBLE is, as it states in the files comments, auto-generated through the NimBLE tool based on NimBLE's `syscfg.yml` files. This ensures, at least for the NimBLE part, that all additions/changes to the default NimBLE configuration are automatically also pulled in RIOT if the NimBLE version is updated. Do you think we could come up with some tooling that for example would take the NimBLE generated file as input and simply remove all the options that are covered by the `mynewt-core` module?

Yes my current solution is of course not ideal. When I ported `uwb-core` I looked into all the options and only kept the ones required, to remove all mynewt-required ones. I'll try to think on different approaches and come back on this.

> 3. At least for the NimBLE package I strongly want to go without patches, as I personally think they are a maintenance nightmare. Gettiing changes into NimBLE code upstream is no big deal, so adapting whats needed adaption there would be the way to go. But I don't quite have good idea on how to upstream the fix to `nimble_porting.c` in a way, that would allow for both, compiling with the NimBLE NPL or the `mynewt-core`. Probably a RIOT specific configuration option in NimBLE would do the trick?!

On the nimble side this can of course be done, I'll adapt accordingly.

File: pkg/mynewt-core/patches/0001-riot-add-interested-header-symlinks.patch
Comment (fjmolinas) [reply pkg/mynewt-core/patches/0001-riot-add-interested-header-symlinks.patch:7]: I have the following options:

1. one patch that similink all interested headers it would look like https://github.com/apache/mynewt-core/commit/8f03dc5f753986d815ef8cea273b19f62f8b7c78
1. add headers under `pkg/mynewt-core/...`
1. add directories to `INCLUDES` with `CFLAGS+=-DHEADER_GUARD` for all headers that we do not want to include, also if there is a name conflict our header must be included first.

From a maintenance perspective I prefer 1 since it requires the fewest modifications at the price of a very simple patch. Next I would prefer 3 but it means adding 20+ CFLAGS and hoping the `INCLUDES` order never gets changed. Do you see other cons with 1? (note that in the current patch I seem to have copied instead of symlink most of those files)

**Comment by fjmolinas** [comment]
>> 1. IMHO it would be great, at least for a transition time, to be able to build NimBLE with both, the NimBLE-included NPL implementation and the new `mynewt-core` package. This would significantly help for bug hunting as we have some means to find out if a faulty behavior is caused by the `mynewt-core` package or if the issue is in the NimBLE code. As discussed offline, I have spend the good part of a year to hunt down bugs in NimBLE and the RIOT NPL implementation, so my upmost concern is to keep the stability we have currently for any means! So this option for regression testing is quite important to me.

>That makes sense, I'll take a look at this.

Done in the latest fixups.

**Comment by fjmolinas** [comment]
> On the nimble side this can of course be done, I'll adapt accordingly.

I changed the patch, it can now be directly upstreamed without breaking anything.

File: pkg/mynewt-core/include/os/os.h
Comment (fjmolinas) [reply pkg/mynewt-core/include/os/os.h:16]: I'll change the license header in most cases and add a change notice when it applies. But in many cases I don't think it applies since its just interface definitions.

For some header definitions, I could simply patch the ones in mynewt-core as well:

```
kernel/os/include/os/os_callout.h
kernel/os/include/os/os_mutex.h
kernel/os/include/os/os_sem.h
kernel/os/include/os/os_task.h
```

But since you prefer to avoid patches should I keep them here?

Comment (fjmolinas) [reply pkg/mynewt-core/include/os/os.h]: I addressed some of them, let me know what you prefer for the others.

File: pkg/mynewt-core/include/syscfg/syscfg.h
Comment (fjmolinas) [reply pkg/mynewt-core/include/syscfg/syscfg.h]: true done.

**Comment by fjmolinas** [comment]
> Yes my current solution is of course not ideal. When I ported uwb-core I looked into all the options and only kept the ones required, to remove all mynewt-required ones. I'll try to think on different approaches and come back on this.

I found out that the easiest was to include the autogenerated `syscfg` from nimble directly, currently its included via a path but this commit can be upstreamed if you agree.

**Comment by fjmolinas** [comment]
@haukepetersen I have addressed most of your comments by either a fix or a question waiting for feedback, those that are waiting for feedback are:

- https://github.com/RIOT-OS/RIOT/pull/16348#discussion_r640526343
- https://github.com/RIOT-OS/RIOT/pull/16348#discussion_r640762518

File: pkg/mynewt-core/include/os/os.h
Comment (fjmolinas) [reply pkg/mynewt-core/include/os/os.h]: https://github.com/RIOT-OS/RIOT/pull/16348/commits/6dc1478aff611c6617746a4bde395f41b813d3bd implements that idea, but I can revert it.

File: pkg/mynewt-core/patches/0001-riot-add-interested-header-symlinks.patch
Comment (fjmolinas) [reply pkg/mynewt-core/patches/0001-riot-add-interested-header-symlinks.patch]: @kaspar030 suggested that I should just use my fork as the reference so to not have to go through patches. That is also an option.

Comment (haukepetersen) [reply pkg/mynewt-core/patches/0001-riot-add-interested-header-symlinks.patch]: 1. and 3. I really don't like: symlinks are IMO quite error prone, and everything depending on the order of include paths will (at least with our current build system) explode at some point and will be quite nasty to debug...
2. although this could easily lead to a situation where the copied header get out of sync with their upstream versions, this seems to be slightly cleaner to me...

A fork seems to be a nice option to me. The benefit I see, is that this would make it much easier for us to have a plain view on the files actually used. But this would still leave the problem in keeping the fork and the upstream mynewt repo in sync...

Aside from those options, I think I have another one we could consider: How about simply deleting the headers we are not using from the include dir? This would be a 2 liner in the Makefile and if we put the list of files to delete in a simple variable these two lines would also very easy to maintain, right?!

File: pkg/mynewt-core/include/os/os.h
Comment (haukepetersen) [reply pkg/mynewt-core/include/os/os.h]: In that case, I think patching would be the safer bet. I would rather take this (potentially wide) detour around potential license issues than avoid patches in any case here. Better safe than sorry :-)

**Comment by fjmolinas** [comment]
Murdock made me realize that now uwb-core was nrf% dependent. Adapted in latest commits.

**Comment by fjmolinas** [comment]
Will rebase to get latest ztimer changes in.

**Comment by fjmolinas** [comment]
> Will rebase to get latest ztimer changes in.

All fixups where pushed independently though.

File: pkg/mynewt-core/patches/0001-riot-add-interested-header-symlinks.patch
Comment (fjmolinas) [reply pkg/mynewt-core/patches/0001-riot-add-interested-header-symlinks.patch]: Did as you suggested!

**Review by MrKevinWeiss — CHANGES_REQUESTED** [review]
I think it would be best to wait until after the release, feel free to dismiss my comment if you disagree or if the release is complete.

**Comment by fjmolinas** [comment]
> I think it would be best to wait until after the release, feel free to dismiss my comment if you disagree or if the release is complete.

This PR as is does not impact nimble, since it also keeps the current workflow as is. Regarding `uwb-core` I'm using this branch since a while now for local work, and since its only a pkg, I think the change is quite contained. I'm not sure this will get in because of the timing, but personally, I don't see it as a hard impact feature. But if you think other work needs more focus I'm fine with it.

**Comment by fjmolinas** [comment]
@haukepetersen do you mind if I squash?

**Comment by MrKevinWeiss** [comment]
> I think the change is quite contained

I didn't look into it, I just saw the lines and files changed and thought it might be a bit much for this release.

**Comment by haukepetersen** [comment]
No, please go ahead and squash, right now it looks good to me.

**Comment by fjmolinas** [comment]
Squashed, there is one patch remaining in Nimnle, haven't opened a PR yet since @miri64 had a comment on the pkg naming. Personally, I still prefer the use the repo name as the pkg name, but I can rename it everywhere to `mynewt-adapter` or another suggestion if you insist (or if other people agree).

File: examples/twr_aloha/Makefile
Comment (kaspar030) [inline examples/twr_aloha/Makefile:56]: is this unrelated?

File: pkg/mynewt-core/Makefile.dep
Comment (kaspar030) [inline pkg/mynewt-core/Makefile.dep:17]: newline missing

File: pkg/mynewt-core/contrib/callout.c
Comment (kaspar030) [inline pkg/mynewt-core/contrib/callout.c:14]: ")"

File: pkg/mynewt-core/contrib/core.c
Comment (kaspar030) [inline pkg/mynewt-core/contrib/core.c:33]: should this have an error case, or is this supposed to be called on anything not covered?

File: pkg/mynewt-core/contrib/cputime.c
Comment (kaspar030) [inline pkg/mynewt-core/contrib/cputime.c:56]: cputime is is supposed to be absolute, right? not sure this if is correct then

File: pkg/mynewt-core/include/os/os_time.h
Comment (kaspar030) [inline pkg/mynewt-core/include/os/os_time.h:14]: ")" ?

File: sys/include/byteorder.h
Comment (kaspar030) [inline sys/include/byteorder.h:30]: is there a way to not include this here? feels like bad style to have a special case for a package.

File: pkg/mynewt-core/contrib/callout.c
Comment (kaspar030) [inline pkg/mynewt-core/contrib/callout.c:2]: some copyrights are "2020", is that up to date?

Comment (fjmolinas) [reply pkg/mynewt-core/contrib/callout.c:2]: Just moved the c file around, I would say the same copyright applies

File: pkg/mynewt-core/contrib/cputime.c
Comment (fjmolinas) [inline pkg/mynewt-core/contrib/cputime.c:14]:
```suggestion
 * @brief       cputime implementation for non nrf5* BOARDs
```

File: pkg/mynewt-core/contrib/cputime.c
Comment (fjmolinas) [reply pkg/mynewt-core/contrib/cputime.c:56]: In case of an overflow?

File: pkg/mynewt-core/contrib/cputime.c
Comment (fjmolinas) [inline pkg/mynewt-core/contrib/cputime.c:67]:
```suggestion
    ztimer_set(ZTIMER_MSEC_BASE, &timer->timer, usecs / US_PER_MS);
```

File: pkg/mynewt-core/contrib/cputime.c
Comment (fjmolinas) [reply pkg/mynewt-core/contrib/cputime.c:56]: TBH this is untested, I don't have a non `nrf5x` BOARD with dw1000 hardware, or nimble support...

File: sys/include/byteorder.h
Comment (fjmolinas) [reply sys/include/byteorder.h:30]: I can see if it can be patched in `mynewt-core`

**Comment by fjmolinas** [comment]
> Squashed, there is one patch remaining in Nimble

Ahh I'm actually wrong here... I found an issue with my `syscfg` workaround :(

**Comment by fjmolinas** [comment]
Comments addressed, I opened PR's for both nimble patches, @haukepetersen is it Ok to merge with the patches while upstream gets merged?

**Comment by kaspar030** [comment]
please squash!

**Review by kaspar030 — APPROVED** [review]
ACK.

Let's wait for a comment from @haukepetersen!

File: pkg/nimble/patches/0001-porting-nimble-src-nimble_port-riot-initializes-time.patch
Comment (haukepetersen) [inline pkg/nimble/patches/0001-porting-nimble-src-nimble_port-riot-initializes-time.patch:18]: Just a thought, and I don't expect this to be fixed in this PR but maybe a follow up later: How about we exclude these two lines for all RIOT builds (e.g. -> `#ifndef RIOT_VERSION`) and call them if needed in the RIOT code in `nimble_riot_init()`(`pkg/nimble/contrib/nimble_riot.c`)?! This way we can get rid of the patch...

**Comment by fjmolinas** [comment]
I'll give this one final round of testing once @haukepetersen takes a look :)

**Review by haukepetersen — APPROVED** [review]
Just rebased my experiment branch on this code and started a 15-BLE-node experiment including the changes from this PR -> everything seems to be still in order. Same building vanilla `gnrc_networking` with NimBLE and this PR yields the expected results, so everything that I can see is working.

I have just the one idea for improvement, but from my side this does not have to be fixed before merging this PR.

ACK, and nice work by the way!

**Comment by fjmolinas** [comment]
Just for the record, this is an application (using this PR as the base) where nodes use BLE (nimble advertisements) for discovery and then based on that initiate TWR acitivity between each other (there are also scanning for current epoch to stay in sync):

- node 1:

```
2021-07-07 11:10:51,488 # main(): This is RIOT! (Version: 2021.07-devel-450-g101a0-pepper/nimble-uwb-core)
2021-07-07 11:10:51,489 # [pepper]: delay for 0s to align uwb_epoch start
2021-07-07 11:10:51,490 # [pepper]: new uwb_epoch t=0
2021-07-07 11:10:51,490 # [pepper]: starting new uwb_epoch
2021-07-07 11:10:52,402 # [pepper]: local ebid: [37, 138, 96, 209, 224, 202, 20, 169, 60, 174, 199, 160, 67, 29, 252, 152, 216, 101, 182, 185, 170, 44, 125, 247, 103, 52, 106, 254, 103, 93, 57, 11, ]
2021-07-07 11:10:52,404 # [pepper]: start adv
2021-07-07 11:10:52,406 # [pepper]: start scanning
> 2021-07-07 11:10:53,274 # [pepper]: time was set back too much, bootstrap from 0
2021-07-07 11:10:53,278 # [pepper]: delay for 23s to align uwb_epoch start
2021-07-07 11:11:16,279 # [pepper]: new uwb_epoch t=25
2021-07-07 11:11:16,282 # [pepper]: starting new uwb_epoch
2021-07-07 11:11:18,049 # [pepper]: local ebid: [49, 48, 240, 243, 220, 27, 116, 121, 254, 237, 12, 121, 180, 215, 57, 242, 82, 156, 117, 130, 87, 251, 253, 129, 147, 195, 182, 78, 169, 60, 124, 25, ]
2021-07-07 11:11:18,051 # [pepper]: start adv
2021-07-07 11:11:18,053 # [pepper]: start scanning
2021-07-07 11:11:21,117 # [uwb_ed]: saw ebid: [181, 206, 217, 51, 199, 10, 149, 26, 180, 211, 217, 143, 50, 53, 228, 128, 104, 135, 38, 98, 62, 153, 222, 131, 214, 199, 126, 107, 42, 41, 112, 57, ]
2021-07-07 11:11:22,978 # [pepper]: 0x4774 at 2cm
2021-07-07 11:11:24,449 # [pepper]: 0x4774 at 8cm
2021-07-07 11:11:27,450 # [pepper]: 0x4774 at 10cm
2021-07-07 11:11:29,984 # [pepper]: 0x4774 at 4cm
2021-07-07 11:11:32,449 # [pepper]: 0x4774 at 8cm
2021-07-07 11:11:34,453 # [pepper]: 0x4774 at 6cm
2021-07-07 11:11:34,984 # [pepper]: 0x4774 at 4cm
2021-07-07 11:11:35,452 # [pepper]: 0x4774 at 7cm
2021-07-07 11:11:36,982 # [pepper]: 0x4774 at 8cm
2021-07-07 11:11:38,454 # [pepper]: 0x4774 at 7cm
2021-07-07 11:11:39,980 # [pepper]: 0x4774 at 33cm
2021-07-07 11:11:40,454 # [pepper]: 0x4774 at 48cm
2021-07-07 11:11:40,980 # [pepper]: 0x4774 at 41cm

```

- node 2:

```
2021-07-07 11:10:51,481 # [mynewt-core]: starting thread uwb_irq
2021-07-07 11:10:51,483 # main(): This is RIOT! (Version: 2021.07-devel-450-g101a0-pepper/nimble-uwb-core)
2021-07-07 11:10:51,485 # [pepper]: delay for 0s to align uwb_epoch start
2021-07-07 11:10:51,486 # [pepper]: new uwb_epoch t=0
2021-07-07 11:10:51,487 # [pepper]: starting new uwb_epoch
2021-07-07 11:10:52,445 # [pepper]: local ebid: [56, 0, 25, 64, 102, 231, 105, 15, 6, 123, 146, 107, 161, 172, 22, 188, 238, 27, 206, 90, 52, 176, 217, 212, 26, 219, 120, 196, 185, 111, 176, 122, ]
2021-07-07 11:10:52,446 # [pepper]: start adv
2021-07-07 11:10:52,449 # [pepper]: start scanning
> 2021-07-07 11:10:53,273 # [pepper]: time was set back too much, bootstrap from 0
2021-07-07 11:10:53,277 # [pepper]: delay for 23s to align uwb_epoch start
2021-07-07 11:11:16,279 # [pepper]: new uwb_epoch t=25
2021-07-07 11:11:16,282 # [pepper]: starting new uwb_epoch
2021-07-07 11:11:18,096 # [pepper]: local ebid: [181, 206, 217, 51, 199, 10, 149, 26, 180, 211, 217, 143, 50, 53, 228, 128, 104, 135, 38, 98, 62, 153, 222, 131, 214, 199, 126, 107, 42, 41, 112, 57, ]
2021-07-07 11:11:18,097 # [pepper]: start adv
2021-07-07 11:11:18,100 # [pepper]: start scanning
2021-07-07 11:11:21,069 # [uwb_ed]: saw ebid: [49, 48, 240, 243, 220, 27, 116, 121, 254, 237, 12, 121, 180, 215, 57, 242, 82, 156, 117, 130, 87, 251, 253, 129, 147, 195, 182, 78, 169, 60, 124, 25, ]
2021-07-07 11:11:22,978 # [pepper]: 0x86ac at 2cm
2021-07-07 11:11:24,449 # [pepper]: 0x86ac at 8cm
2021-07-07 11:11:27,450 # [pepper]: 0x86ac at 10cm
2021-07-07 11:11:29,984 # [pepper]: 0x86ac at 4cm
2021-07-07 11:11:32,449 # [pepper]: 0x86ac at 8cm
2021-07-07 11:11:34,453 # [pepper]: 0x86ac at 6cm
2021-07-07 11:11:34,984 # [pepper]: 0x86ac at 4cm
2021-07-07 11:11:35,452 # [pepper]: 0x86ac at 7cm
2021-07-07 11:11:36,982 # [pepper]: 0x86ac at 8cm
2021-07-07 11:11:38,454 # [pepper]: 0x86ac at 7cm
2021-07-07 11:11:39,980 # [pepper]: 0x86ac at 33cm
2021-07-07 11:11:40,454 # [pepper]: 0x86ac at 48cm
2021-07-07 11:11:40,980 # [pepper]: 0x86ac at 41cm
2021-07-07 11:11:41,454 # [pepper]: 0x86ac at 65cm
2021-07-07 11:11:41,984 # [pepper]: 0x86ac at 76cm
2021-07-07 11:11:42,454 # [pepper]: 0x86ac at 70cm
2021-07-07 11:11:44,985 # [pepper]: 0x86ac at 80cm
2021-07-07 11:11:45,985 # [pepper]: 0x86ac at 74cm
```

**Comment by fjmolinas** [comment]
Thanks for the feedback and review @haukepetersen @kaspar030!

**Pull request merged**
